rmw_cyclonedds_cpp/src/rmw_node.cpp

```diff
@@ -1929,7 +1929,6 @@ static CddsPublisher * create_cdds_publisher(
   dds_delete_qos(qos);
 fail_qos:
   dds_delete(topic);
-  ddsi_sertype_unref(stact);
```

Comment:
There's a few things in this function that are unclear to me:

1.  Why do we need the "stact" variable at all?  As far as I can tell, if it is provided to `create_topic`, it just saves a copy of the `sertype` pointer.  But we already have that pointer (we are passing it into `create_topic` to begin with).  So why do we need it?

2.  If `create_topic` fails for some reason, I feel like we are leaking the `sertype` variable; `create_sertype` allocates heap space for it, but nothing frees it in the error case.

3.  Is `dds_delete` guaranteed to free `sertype` in the case where `create_topic` succeeds, but `create_readwrite_qos` fails?  (I'm guess this is the case, given the nature of this bug, but I'm just checking).

Reply:
> So why do we need it?

There is a bit of trickery going on: one constructs a `sertype` and passes it to `create_topic` but it isn't a given that this is the one that will actually be used. Instead, `dds_create_topic_sertype` checks whether it has an equivalent one already, and if it has one, uses that instead. It always consumes the reference, which means that in the latter case it drops the `sertype` (because it was just created and there is no other reference).

The pointer it returns is a "raw" pointer, so there is no unref'ing of that (unless you increment the refcount yourself). Its lifetime is *at least* that of the topic object and of any readers/writers that you created with that topic object. The latter bit is why calling `dds_delete(topic)` in the successful path of creating a publisher is fine: it all remains valid because the writer still exists.

(You could increment the refcount prior to calling `dds_create_topic_sertype`, that would keep it alive, but it would mean you might be using a different `sertype` than the DDS writer is using. That in turn means some relatively expensive conversions are required when you try to publish a serialised message. So it makes more sense to do this.)

> If create_topic fails for some reason [...]

That case is handled inside `create_topic` in `rmw_node.cpp`: if `dds_create_topic_sertype` fails, `create_topic` immediately drops the `sertype` it just created. If it succeeds, it either retains the `sertype` or drops it to use an already existing one instead. In the former case, it is freed immediately (because there aren't any other references around); in the latter, it will be freed whenever the last reference disappears.

> Is dds_delete guaranteed to free sertype in the case where create_topic succeeds, but create_readwrite_qos fails?

Yes, at that point `topic` owns the reference. (Well, in the simplified version in which there are no other threads creating topics using an equivalent `sertype` in parallel anyway ...)